gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/DebouncingAsyncFileListener.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.versions.intellij;
+
+import com.intellij.openapi.Disposable;
+import com.intellij.openapi.vfs.AsyncFileListener;
+import com.intellij.openapi.vfs.newvfs.events.VFileEvent;
+import com.intellij.util.Alarm;
+import com.intellij.util.SingleAlarm;
+import java.util.ArrayList;
+import java.util.List;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class DebouncingAsyncFileListener implements AsyncFileListener, Disposable {
+    private static final Logger log = LoggerFactory.getLogger(DebouncingAsyncFileListener.class);
+
+    private final AsyncFileListener delegate;
```

Comment:
In general, if locking an object using the `synchronized` keyword it's best to [annotate the field with `@GuardedBy` from errorprone](https://errorprone.info/bugpattern/GuardedBy). This will ensure that any usages of the field are locked appropriately. It's ever so easy to modify the class and forget to do this (and suddenly everything breaks in the worse possible way).

But it's also generally better to use some type that intrinsically thread safe, which `ArrayList` is not. The best option here is likely a [`BlockingQueue`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/BlockingQueue.html) implementation (`BlockingQueue` is fantastic and generally the go to primitive when dealing with concurrent code). It has a [`drainTo` ](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/BlockingQueue.html#drainTo(java.util.Collection):~:text=int%C2%A0drainTo,E%3E%C2%A0c))method which allows you to just grab all of the elements from the queue in a thread safe way. Of the implementations, `ArrayBlockingQueue` is more performant, but it has a fixed capacity and we don't know how many elements there will be, so I'd use a `LinkedBlockingQueue`.

Reply:
(I'm assuming here `SingleAlarm` is threadsafe and doesn't need locking).

Reply:
I've switched to using `LinkedBlockingQueue` and have added a general `FilteringAsyncFileListener` which is used to filter out all but props and locks changes